Add Cupertino to gallery and add CupertinoButton and many yak friends

xster opened the pull request:
#8352

xster committed:
Add cupertino to gallery and add CupertinoButto

xster commented:
cc @abarth 
It's not ready yet. I just wanted to have some code so I can ask you some concrete questions. There's some animation stuff I didn't understand. #8409

Hixie commented on examples/flutter_gallery/lib/demo/buttons_demo.dart:
```diff
@@ -37,6 +38,10 @@ const String _actionText =
     "point.";
 
+const String _iosText =
+    "iOS themed buttons are flat. They can have borders or backgrounds but only "
```
nit: style says to use single quotes

xster replied:
Done

Hixie commented on examples/flutter_gallery/lib/demo/buttons_demo.dart:
```diff
@@ -2,6 +2,7 @@
 // Use of this source code is governed by a BSD-style license that can be
 // found in the LICENSE file.
```
we should probably not include Cupertino widgets long-term as a separate demo. Instead #8410 should take care of that (since you can switch the OS from the drawer in the gallery, or from hot reload).

abarth replied:
The long term plan is somewhat hazy.  The idea is to put these in the gallery under a separate heading and then change the presentation in the gallery once the overall structure becomes clearer.

sethladd replied:
And, we're exploring Gallery design ideas specifically for iOS-flavor of gallery. But that's not an immediate thing.

xster replied:
Ya, this is not a long term solution. I think the remaining thing to figure out still with #8410 is how to treat more complex things that are part of the information architecture like the scaffold and app bar and how they fit in the gallery.

Hixie commented on examples/flutter_gallery/lib/demo/buttons_demo.dart:
```diff
@@ -246,4 +257,21 @@ class _ButtonsDemoState extends State<ButtonsDemo> {
     );
   }
+
+  Widget buildIosButton() {
```
The capitalisation should be "iOS". If that doesn't fit, make it private and use "Cupertino".

xster replied:
Done

xster committed:
Use single quotes

Hixie commented on examples/flutter_gallery/lib/demo/buttons_demo.dart:
```diff
@@ -246,4 +257,21 @@ class _ButtonsDemoState extends State<ButtonsDemo> {
+          new CupertinoButton(
+            text: 'iOS Button',
+            onPressed: () {
+              // Perform some action
```
if you do end up keeping this, make this actually do something (like pop up a snackbar)

xster replied:
The existing demo doesn't do anything on press. Might be a bit annoying if things kept popping up on each press for each tap vs a dialog result etc

Hixie commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,116 @@
+class CupertinoButton extends StatefulWidget {
+  CupertinoButton({
+    this.text,
+    this.padding = 16.0,
```
colon

xster replied:
Done

Hixie commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,116 @@
+    return new GestureDetector(
+      onTapDown: config.enabled ? _handleTapDown : null,
+      onTapUp: config.enabled ? _handleTapUp : null,
+      onTap: config.enabled ? _handleTap : null,
```
why not just `onTap: config.onPressed` ?

xster replied:
Done

Hixie commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,116 @@
+  @override
+  void initState() {
+    super.initState();
+    _animationController = new AnimationController(
```
don't forget to dispose the controlled in your dispose method

xster replied:
Done

Hixie commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,116 @@
+        constraints: new BoxConstraints(minWidth: 48.0, minHeight: 48.0),
+        child: new Padding(
+          padding: new EdgeInsets.all(config.padding),
+          child: new AnimatedDefaultTextStyle(
```
why AnimatedDefaultTextStyle since the style is a constant?

xster replied:
Done

xster commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,116 @@
+  }
+
+  void _handleTapDown(TapDownDetails details) {
+    _animationController.animateTo(0.2);
```
Not sure if I'm doing something wrong. Even on release builds, it feels pretty sluggish. Should I be using something else than the GestureDetector for instantaneous UI feedback?

abarth replied:
Tap down is delayed by 100 ms.  If you prefer, you can use a `Listener` and listen to raw pointer events, which are not delayed.

xster replied:
Ah, thanks. Done.

The native button has a curve on the way in too but it's faster. Tweaked the numbers. I'm pretty happy with the effects.

xster committed:
Add disabled state

abarth commented on examples/flutter_gallery/lib/demo/buttons_demo.dart:
```diff
@@ -80,6 +85,12 @@ class _ButtonsDemoState extends State<ButtonsDemo> {
         exampleCodeTag: _actionCode,
       ),
+      new ComponentDemoTabData(
+        tabName: 'iOS',
```
I'd probably call it Cupertino

xster replied:
We want to label the public facing parts of this widget set Cupertino too?

abarth replied:
For now, yes.  We'll probably change it later as we rationalize things more, but Cupertino is a good place to start.

xster replied:
Done

abarth commented on examples/flutter_gallery/lib/demo/buttons_demo.dart:
```diff
@@ -80,6 +85,12 @@ class _ButtonsDemoState extends State<ButtonsDemo> {
       ),
+      new ComponentDemoTabData(
+        tabName: 'iOS',
+        description: _iosText,
```
IMHO, we should use "cupertino" consistently in the code when referring to these widgets.

xster replied:
Done

abarth commented on examples/flutter_gallery/lib/demo/dialog_demo.dart:
```diff
@@ -195,7 +196,51 @@ class DialogDemoState extends State<DialogDemo> {
+          new RaisedButton(
+            child: new Text('iOS ALERT'),
+            onPressed: () {
+              showDemoDialog<DialogDemoAction>(
```
Cupertino dialogs should have a non-dismissible barrier.

xster replied:
Done

abarth commented on examples/flutter_gallery/lib/demo/progress_indicator_demo.dart:
```diff
@@ -73,6 +74,7 @@ class _ProgressIndicatorDemoState extends State<ProgressIndicatorDemo> with Sing
       new Row(
         mainAxisAlignment: MainAxisAlignment.spaceEvenly,
         children: <Widget>[
+          new CupertinoActivityIndicator(),
```
Rather than putting these in the same demos as the material widgets, I think we should create a separate heading in the home page of the gallery and show the widgets separately.

xster replied:
I think we probably need a long term solution for cupertino and gallery though for the short term, I think it's good to have an easy comparison between different widget skins of the same widget 'type'.

Might be confusing for the user if we sliced both horizontally with widget 'function categories' and vertically with 'widget look/use circumstance'.

abarth replied:
We can talk about this more in person if you like, but for now, I think it will be easier if we put the cupertino widgets in their own section on the home screen and we don't mix cupertino and non-cupertino in the same demo.

xster replied:
Done

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,113 @@
+    @required this.onPressed,
+  });
+
+  final String text;
```
Rather than taking text, this should take a child widget.  The client can supply a Text widget if they want, but they can also supply something else more interesting.

xster replied:
Done

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,113 @@
+
+  final String text;
+
+  final double padding;
```
Padding should be an `EdgeInsets`

xster replied:
Oh, makes sense. Done

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  @override
+  Widget build(BuildContext context) {
```
We usually put the `build` function after the methods it references.

xster replied:
Done

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,113 @@
+    return new GestureDetector(
+      onTapDown: config.enabled ? _handleTapDown : null,
+      onTapUp: config.enabled ? _handleTapUp : null,
+      onTapCancel: config.enabled ? _handleTapCancel : null,
```
I'd pull `config.enabled` into a local variable rather than calling this getter three times.

xster replied:
Done

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,113 @@
+            ),
+            child: new Text(
+              config.text,
+              style: config.enabled ? _kCupertinoButtonStyle : _kCupertinoDisabledButtonStyle,
```
When you take a `child` widget, you can use `DefaultTextStyle` to control the style of Text widgets inside the child.

xster replied:
Ya, noticed that in the other widgets. Just curious though what the anticipated use case is. What about users that would put for instance something complex like a table as a child of the button?

xster replied:
Takes a child now

xster committed:
Merge branch 'master' of github.com:flutter/flutter into cupertino-button

xster committed:
Some review notes

xster committed:
Make button animation more responsive and tweak timing

xster committed:
Renamed things Cupertino

xster committed:
Button with background, move cupertino demos, move material demos

xster committed:
Move 2 level list too

xster committed:
Merge branch 'master' of github.com:flutter/flutter into cupertino-button

# Conflicts:
#	examples/flutter_gallery/lib/demo/all.dart

xster commented:
Grew a yak farm in the meantime. PTAL
- Added a cupertino button version with background
- Tweaked the cupertino dialog layout a bit according to http://ios10.greatsimple.io/
- Extracted the various cupertino demo to their own pages
- Rearranged the flutter_gallery/lib/demo folder a bit
- Let cupertino button take a child

abarth commented on examples/flutter_gallery/lib/demo/cupertino/cupertino_activity_indicator_demo.dart:
```diff
@@ -0,0 +1,22 @@
+import 'package:flutter/material.dart';
+
+class CupertinoProgressIndicatorDemo extends StatelessWidget {
+  static const String routeName = '/cupertino_progress_indicator';
```
Maybe `/cupertino/progress_indicator` ?  These names don't really matter, but might as well be pretty.

xster replied:
SG. Changed all the components demo routes to [cupertino/material]/[name]

abarth commented on examples/flutter_gallery/lib/demo/cupertino/cupertino_buttons_demo.dart:
```diff
@@ -0,0 +1,67 @@
+                      new CupertinoButton(
+                        child: new Text('Cupertino Button'),
+                        onPressed: () {
+                          // Perform some action
```
Maybe show a snackbar so there's some feedback?

xster replied:
@Hixie was saying that too :) 

The snackbar is a little slow for these fast interaction components. Added a text counter.

abarth commented on packages/flutter/lib/cupertino.dart:
```diff
@@ -8,7 +8,9 @@
 export 'src/cupertino/slider.dart';
 export 'src/cupertino/switch.dart';
 export 'src/cupertino/thumb_painter.dart';
+export 'widgets.dart';
```
I'm not sure about this line.  It's true we do this in material.dart, which is a good reason to do it here.  However, this will make the dartdoc for Cupertino pretty overwhelming...  I'm inclined to say we should skip this for now.  We can always add it later, but once we add it, it's hard to remove because removing it will break clients.

xster replied:
Ah, didn't know it affects dartdoc. Removed.

xster replied:
Ah, didn't know it affects dartdocs. Removed

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,167 @@
+  static const Color kCupertinoBlue = const Color(0xFF007AFF);
+  static const Color kCupertinoWhite = const Color(0xFFFFFFFF);
+  static const Color kCupertinoDisabledBackground = const Color(0xFFA9A9A9);
+  static const Color kCupertinoDisabledForeground = const Color(0xFFC4C4C4);
```
No need to repeat the name "cupertino" here.  It's already in the class name.

xster replied:
Done

xster replied:
Done

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,167 @@
+///
+///  * <https://developer.apple.com/ios/human-interface-guidelines/ui-controls/buttons/>
+class CupertinoButton extends StatefulWidget {
+  // TODO(xster): move this to a common Cupertino color palatte with the next yak.
```
Yeah, we should have a CupetinoTheme or something.

xster replied:
Will do in next yak farm

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,167 @@
+
+  static const EdgeInsets _kCupertinoButtonPadding = const EdgeInsets.all(16.0);
+  static const EdgeInsets _kCupertinoBackgroundButtonPadding =
+      const EdgeInsets.symmetric(vertical: 16.0, horizontal: 64.0);
```
For private constants, we usually just declare them at the file level.  It doesn't make any difference one way or another.  That's just the convention we've been using.

xster replied:
If I understand right, doesn't having them at the file level make them global (library) fields once the file is imported? 

Don't really have a strong opinion but maybe once we have 20 composable cupertino files, it might be annoying for autocompleters to give _kButtonInnerPadding and _kPrimaryTextFontSize etc that you imported as soon as you type '_k' when you're in CupertinoScaffold and want something qualified like CupertinoPicker._kHeight

Hixie replied:
That won't be a problem because autocomplete won't complete privates across files, since they're per-file private.

xster replied:
Oh ya, you're right, it doesn't. Neat. Moved.

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,167 @@
+    @required this.onPressed,
+  });
+
+  final Widget child;
```
There's a boilerplate dartdoc we use for this.  You can see it on RaisedButton or whatever.

xster replied:
Done

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,167 @@
+}
+
+class _CupertinoButtonState extends State<CupertinoButton> with SingleTickerProviderStateMixin {
+  static const Duration kFadeOutDuration = const Duration(milliseconds: 10);
```
You might comment where you got these values from.

xster replied:
Done

abarth commented on packages/flutter/lib/src/cupertino/button.dart:
```diff
@@ -0,0 +1,167 @@
+      child: new GestureDetector(
+        onTap: config.onPressed,
+        child: new ConstrainedBox(
+          constraints: const BoxConstraints(minWidth: 48.0, minHeight: 48.0),
```
If you just put this widget in a Center inside a Scaffold body, I think you'll end up with something super tall.  You might want to set the maxHeight as well.

xster replied:
Ya, I had this problem with width. Seems like the width/heightFactors below gets the intended effect

abarth replied:
Looks good.

abarth commented:
![LGTM](http://i3.cpcache.com/product/1351564863/lgtm_bumper_stickers.jpg?width=200&height=200)

abarth commented:
Please add a basic control test for CupertinoButton

xster committed:
Refactor various demo route names

xster committed:
Merge branch 'master' of github.com:flutter/flutter into cupertino-button

# Conflicts:
#	examples/flutter_gallery/lib/demo/all.dart

xster committed:
Some review notes

xster committed:
More reviews and add test

xster commented:
Oops, forgot test. Added

xster committed:
Linter as

xster committed:
Move private constant up

abarth commented on packages/flutter/test/cupertino/button_test.dart:
```diff
@@ -0,0 +1,101 @@
+    RenderBox buttonBox = tester.renderObject(find.byType(CupertinoButton));
+    expect(
+      buttonBox.size,
+      greaterThanOrEqualTo(const Size.square(48.0)),
```
Why not just put in the exact size?

abarth replied:
If you're worried about text metrics, I'd just put a SizedBox as the child instead of text.  That will give you exact control over what you're testing.

Hixie replied:
Text metrics shouldn't be an issue any more. Ahem font means characters have the exact width and height of the font size.

abarth replied:
Great

xster replied:
Does that mean if I have 2 characters of a non monospace font at size 15, the dimensions are exactly 30x15 logical pixels?

Hixie replied:
If you have two characters in the Ahem font at size 15, then the dimensions will be exactly 30x15, with the baseline at exactly 12 pixels from the top.

(This only works if the characters are in the Ahem font's character set, which covers most ASCII characters including in particular A-Z, a-z, 0-9, and space.)

In addition, the space character will be drawn as a transparent square, the É character will be drawn as fill above the baseline and transparent below, the p character will be drawn as transparent above the baseline and filled below, and the other characters will be drawn as filled squares.

xster replied:
Oh cool, will try. 

Sorry, I didn't realize 'Ahem' was a literal font name.

abarth commented:
Oh, sorry.  I was going to recommend that you replace the text with a SizedBox in your layout tests so your tests aren't sensitive to fickle text metrics.
